datafusion/physical-optimizer/src/enforce_sorting/mod.rs

```diff
@@ -126,29 +126,65 @@ fn update_sort_ctx_children(
 /// [`CoalescePartitionsExec`] descendant(s) for every child of a plan. The data
 /// attribute stores whether the plan is a `CoalescePartitionsExec` or is
 /// connected to a `CoalescePartitionsExec` via its children.
+///
+/// The tracker halts at each [`SortExec`] (where the SPM will act to replace the coalesce).
+///
+/// This requires a bottom-up traversal was previously performed, updating the
+/// children previously.
 pub type PlanWithCorrespondingCoalescePartitions = PlanContext<bool>;
```

Comment:
I did a refactor.
* The `parallelize_sorts` and its helper `remove_bottleneck_in_subplan` removed a bunch of extra nodes, which were then added back in a few conditionals. 
* instead, if I tightened up the context `PlanWithCorrespondingCoalescePartitions` and how it was built in `update_coalesce_ctx_children` -- then we found (a) avoid excess node removal, and also (b) no longer add back needed nodes that were removed.